533 Adds NDAs support for recap email

albertisfu opened the pull request:
Support to parse and download free documents for NDAs was added.

It was necessary to add a new field to support download documents for NDAs `email_notice_type` due to download link structures are different.

Here is some example data for an NDA:
```
{
  "case_name": "New York State Telecommunicati v. James",
  "court_id": "ca2",
  "date_filed": "2022-03-23",
  "docket_entries": [
    {
      "date_filed": "2022-03-23",
      "description": "REPLY BRIEF, on behalf of Appellant Letitia A. James, FILED. Service date 03/23/2022 by CM/ECF. [3283515] [21-1975]",
      "document_number": "00208754210",
      "document_url": "https://ecf.ca2.uscourts.gov/docs1/00208754210?uid=59a7f7615f78153b",
      "pacer_case_id": "21-1975",
      "pacer_doc_id": "00208754210",
      "pacer_magic_num": "59a7f7615f78153b",
      "pacer_seq_no": null
    }
  ],
  "docket_number": "21-1975",
  "email_notice_type": "NDA",
  "email_recipients": [
    {
      "email_addresses": [
        "recipient@recap.email"
      ],
      "name": "Recap email recipient, -:"
    }
  ]
}
``` 
Added tests using examples from the emails we've received from different courts but they aren't too many, so as long as we receive more NDA's it might be necessary to adjust the parser to consider all possible cases. 

Let me know what do you think and if it's ok I can do a version bump and update the CL PR.

mlissner commented on juriscraper/pacer/utils.py:
```diff
@@ -59,9 +60,18 @@ def get_pacer_case_id_from_doc1_url(url):
+    """Extract the magic number from the doc1 URL."""
+    if email_notice_type == "NDA":
+        # NDA free look link format is:
+        # https://ecf.ca2.uscourts.gov/docs1/00208721516?uid=b775e9908ad79ce2
```
Huh. I'm not familiar with the NDAs enough. You feel good about this being the magic number?

albertisfu replied:
Yes, at least within the NDAs examples we have now if there is a document link available it has this format where `uid=b775e9908ad79ce2` is the "magic number".

In fact, if you try to go to this link (anonymously) that it's already used you'll see the following:
[https://ecf.ca2.uscourts.gov/docs1/00208721516?uid=b775e9908ad79ce2](https://ecf.ca2.uscourts.gov/docs1/00208721516?uid=b775e9908ad79ce2)
![Screen Shot 2022-08-12 at 18 09 41](https://user-images.githubusercontent.com/486004/184455511-046f6902-ed00-4ac3-b3f4-ac72fec3f1d2.png)

mlissner reviewed (commented):
One small comment. The rest makes sense and looks good. If you feel good about the comment, please do a release. Don't forget the changelog if there is one (I think there is?)

mlissner commented on juriscraper/pacer/reports.py:
```diff
@@ -176,6 +176,7 @@ def download_pdf(
         pacer_case_id: str,
         pacer_doc_id: int,
         pacer_magic_num: Optional[str] = None,
+        email_notice_type: Optional[str] = None,
```
I think we could simplify this parameter by calling it `appellate` and making it a boolean that's defaulted to False:

```suggestion
        appellate: bool = False,
```

Because I think this applies beyond NDAs vs. NEFs, and as I saw over in CourtListener, it's weird to be saying *anything* about emails when calling this function elsewhere.

albertisfu replied:
Yeah, I agree, I've changed and set it to False as default.

albertisfu commented:
Thanks! I've changed the `email_notice_type` parameter to `appellate`.

Also added the version bump but I think is not ready because I updated my CL dev image with it and ran more tests including some new NDA's we received today and there is a new one from `ca11` that it's failing to parse the document description properly. So I'll be back with the fix it and add more tests.

albertisfu commented:
I've solved the problem parsing the description that I detected in ca11.

Now we also get the pacer_case_id for NDAs
As we discussed on https://github.com/freelawproject/courtlistener/pull/2257 the document_number now is null for NDAs
Added new examples (recipients anonymized)
Let me know if there is anything else. I'll also update CL PR with these changes.

mlissner commented on juriscraper/pacer/reports.py:
```diff
@@ -198,6 +199,11 @@ def download_pdf(
                 "caseid": pacer_case_id,
                 "magic_num": pacer_magic_num,
             }
+            if appellate:
+                url = make_docs1_url(self.court_id, pacer_doc_id)
+                params = {
+                    "uid": pacer_magic_num,
+                }
```
Should there be an `else` here for the normal path, so that the `if appellate` branch doesn't get the `magic_num` and `caseid` params?

albertisfu replied:
In this case, I was overriding the params completely before passing to the request so in case it's appellate only passes the `uid` without the `magic_num` or the` caseid`. But yes I think it's more clear using the `else`, so I added it.

mlissner reviewed (commented):
This looks OK, I think. A few comments, thanks.

johnhawkinson commented on juriscraper/pacer/email.py:
```diff
@@ -227,7 +286,10 @@ def _get_docket_entries(
                 )
-                document_number = self._get_document_number()
+                if self._is_appellate():
+                    document_number = None
```
No. There are appellate _cases_ which have document numbers, but they don't show up in NDAs AFAIK.

mlissner replied:
Ah. Thanks for chiming in.

albertisfu replied:
Yeah, that's right into the NDAs we don't have the document number, that's why is set as None. So in Courtlistener, we'll add it as a blank field.
